sound/soc/codecs/max98373-sdw.c

```diff
@@ -0,0 +1,1203 @@
+		dpn[i].num = bit;
+		dpn[i].type = SDW_DPN_SIMPLE;
+		dpn[i].simple_ch_prep_sm = true;
+		dpn[i].ch_prep_timeout = 10;
+		i++;
+	}
+
+	/* Allocate port_ready based on num_of_ports */
+	slave->port_ready = devm_kcalloc(&slave->dev, num_of_ports,
+					sizeof(*slave->port_ready),
+					GFP_KERNEL);
+	if (!slave->port_ready)
+		return -ENOMEM;
+
+	/* Initialize completion */
+	for (i = 0; i < num_of_ports; i++)
+		init_completion(&slave->port_ready[i]);
+
+	/* set the timeout values */
+	prop->clk_stop_timeout = 20;
+
+	/* wake-up event */
+	prop->wake_capable = 1;
+	return 0;
+}
+
+static int max98373_bus_config(struct sdw_slave *slave,
+			    struct sdw_bus_params *params)
+{
+	int ret;
+
+	ret = max98373_clock_config(slave, params);
+	if (ret < 0)
+		dev_err(&slave->dev, "Invalid clk config");
+
+	return 0;
+}
+
+static int max98373_interrupt_callback(struct sdw_slave *slave,
+				    struct sdw_slave_intr_status *status)
+{
+	pr_debug("%s control_port_stat=%x", __func__, status->control_port);
+	return 0;
+}
+
+/*
+ * slave_ops: callbacks for get_clock_stop_mode, clock_stop and
+ * port_prep are not defined for now
+ */
+static struct sdw_slave_ops max98373_slave_ops = {
+	.read_prop = max98373_read_prop,
+	.interrupt_callback = max98373_interrupt_callback,
+	.update_status = max98373_update_status,
+	.bus_config = max98373_bus_config,
+};
+
+static int max98373_sdw_probe(struct sdw_slave *slave,
+			   const struct sdw_device_id *id)
+{
+	struct regmap *regmap;
+	int ret = 0;
+
+	/* Assign ops */
+	slave->ops = &max98373_slave_ops;
+
+	/* Regmap Initialization */
+	regmap = devm_regmap_init_sdw(slave, &max98373_sdw_regmap);
+	if (!regmap)
+		return -EINVAL;
+
+	max98373_init(slave, regmap);
+
+	return ret;
+}
+
+static int max98373_sdw_remove(struct sdw_slave *slave)
+{
+	max98373_reset(slave);
+	return 0;
+}
+
+#if defined(CONFIG_OF)
+static const struct of_device_id max98373_of_match[] = {
+	{ .compatible = "maxim,max98373", },
+	{},
+};
+MODULE_DEVICE_TABLE(of, max98373_of_match);
+#endif
+
+#ifdef CONFIG_ACPI
+static const struct acpi_device_id max98373_acpi_match[] = {
+	{ "MX98373", 0 },
+	{},
+};
+MODULE_DEVICE_TABLE(acpi, max98373_acpi_match);
+#endif
+
+static const struct sdw_device_id max98373_id[] = {
+	SDW_SLAVE_ENTRY(0x019F, 0x8373, 0),
+	{},
+};
+MODULE_DEVICE_TABLE(sdw, max98373_id);
+
+static struct sdw_driver max98373_sdw_driver = {
+	.driver = {
+			.name = "max98373",
+			.owner = THIS_MODULE,
+			.of_match_table = of_match_ptr(max98373_of_match),
+			.acpi_match_table = ACPI_PTR(max98373_acpi_match),
+			.pm = &max98373_pm,
+		   },
+	.probe = max98373_sdw_probe,
+	.remove = max98373_sdw_remove,
+	.ops = &max98373_slave_ops,
+	.id_table = max98373_id,
+};
+
+static int __init sdw_slave_init(void)
+{
+	return sdw_register_driver(&max98373_sdw_driver);
+}
+
+static void __exit sdw_slave_exit(void)
+{
+	sdw_unregister_driver(&max98373_sdw_driver);
+}
+
+module_init(sdw_slave_init);
+module_exit(sdw_slave_exit);
```

Comment:
module_sdw_driver() does all this for you

Reply:
Agreed. Will replace it. Thanks.